Allow to change the schedule on run-time

pudkrong opened the pull request.

pudkrong committed:
feat: update fetching interval in run time

pudkrong committed:
feat: remove cache on parameter store read

Reading from parameter store is free. So, no need cache.

pudkrong committed:
fix: change e2e test after rebase to saga

pudkrong committed:
refactor: removed unused code

pudkrong committed:
test: make step test more meaningful

pudkrong committed:
refactor: refactor code

pudkrong committed:
fix: merge conflicts

pudkrong committed:
refactor: rename variable

pudkrong committed:
test: increase retry interval

pudkrong committed:
fix: using stack region

pudkrong committed:
Merge branch 'saga' into pudkrong/issue18

pudkrong committed:
build: set max connections to parameter store extension

pudkrong committed:
Merge branch 'saga' into pudkrong/issue18

coderbyheart reviewed (commented):
On hold until we find the issue with the message bridge.

pudkrong committed:
Merge branch 'saga' into pudkrong/issue18

pudkrong committed:
feat: remove duplicated device id when request shadow

pudkrong committed:
fix: all exceptions in handler will be catched and unlock process

pudkrong committed:
test: allow a bit margin for schedule

Since Cloudwatch scheduler has 1 minute lower bound for cron job, then using delayed message in SQS might have a slight margin for cron job

pudkrong committed:
Merge branch 'saga' into pudkrong/issue18

pudkrong committed:
Merge branch 'saga' into pudkrong/issue18

pudkrong committed:
Merge branch 'saga' into pudkrong/issue18

pudkrong commented:
@coderbyheart Can we merge this PR?

coderbyheart commented:
I'll review it today.

coderbyheart commented on websocket/deviceShadowUpdateChecker.ts:
```diff
@@ -0,0 +1,116 @@
+			{ count: Number.MAX_SAFE_INTEGER, interval: 5 },
+		]
+		const modelStep =
+			config[device.model?.replace(/[^\w.-]/g, '_')] ?? defaultStep
```
Use exact model name, otherwise we risk overlapping names. Use base64 to encode model name as parameter name.

pudkrong replied:
Parameter store allows only `a-zA-Z0-9_.-` as per https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html

coderbyheart replied:
Yes, I know.

coderbyheart replied:
If you don't want base64 with additional character representation, you can sha1 hash the model name.

pudkrong replied:
sha1 is good. Base64 contains un-allowed characters

coderbyheart commented:
Please add a CLI command to configure the schedule.

pudkrong committed:
Merge branch 'saga' into pudkrong/issue18

pudkrong committed:
fix: hash model name before saving into SSM

pudkrong committed:
feat: cli to configure shadow fetcher

coderbyheart commented on cli/commands/set-shadow-fetcher.ts:
```diff
@@ -0,0 +1,63 @@
+			system: 'stack',
+		})({
+			property: modelHash,
+			value,
```
Validate input, because this is not a simple value.

coderbyheart commented on util/hashSHA1.ts:
Needs test, if it's a separate module.

pudkrong replied:
done

pudkrong committed:
feat: validate value of shadow fetcher cli

pudkrong committed:
test: add unit test

coderbyheart commented on cli/commands/set-shadow-fetcher.ts:
```diff
@@ -41,6 +42,12 @@ export const setShadowFetcherCommand = ({
 			throw new Error(`Must provide value!`)
 		}
 
+		if (!/^(?:\d+(:\d+)?(\s*,\s*)?)+$/.test(value)) {
+			throw new Error(
+				`Invalid value. It should be in format: <interval> or <interval>:<count> or <interval>:<count>, <interval>:<count>, ...`,
+			)
+		}
```
*nitpick*: allows leading zeros

coderbyheart commented on README.md:
```diff
@@ -81,3 +81,21 @@ Afterwards you can run the simulator using
+- interval:count
+- interval:count, interval:count
+
+Here, `interval` refers to the duration in seconds for fetching the shadow, and
+`count` represents the number of iterations.
```
```suggestion
Here, `interval` refers to the interval in seconds between fetching the shadow of a device, and
`count` represents the number of iterations.
```

Please provide an example, e.g. what would 17:42 mean technically?

pudkrong committed:
docs: give example

pudkrong committed:
fix: not allow leading zeros
